## Data Cleaning and Preparation

<p>Data cleaning and preparation are crucial steps in the data analysis process. These steps involve transforming raw data into a format that is suitable for analysis, and ensuring that the data is accurate, complete, and consistent. </p>

The following are some key steps involved in our data cleaning and preparation process:

>1. **Data Collection**: our raw dataset is from the 2021 BRFSS survey data done by the Center of Disease Control and Prevention based on over 400,000 survey participants in the United States. The original data file can be accessed at: https://www.cdc.gov/brfss/annual_data/annual_2021.html


>2. **Data Extraction**: the survey dataset consists of 303 columns, which contain responses for different questions asked in the survey. To identify factors releted to our research area, i.e., hypertension, extensive research was conducted. After which, relevant variable columns are identified and extracted from the survey dataset.


>3. **Data Cleaning**: numerous steps were taken to clean the dataset, including tackling missing values from survey respondents, dropping irrelevant responses and removing outliers for numeric data.


>4. **Data Transformation**: after cleaning, the data is transformed into a format that is suitable for analysis. This involves creating new variables by manipulating existing columns and decoding categorical data into corresponding levels.


>5. **Data Documentation & Export**: the last step involves a detailed documentation of the data. Our codebook can be found in the data description file.



### Step 0: Convert the 2021 BRFSS data file to csv 

The original survey data file is only available in SAS Transport or ASCII Format. Thus, it has to be converted to csv format to be further processed using Pandas' libraries.

The file conversion is conducted on our local machine through the code below:</br>
`-m xport LLCP2021.xpt > raw.csv`

In [1]:
# Basic Libraries
import numpy as np
import pandas as pd
import seaborn as sb #for graphics
import matplotlib.pyplot as plt # we only need pyplot
sb.set() # set the default Seaborn style for graphics

In [2]:
#Import raw survey dataset
raw_dataset = pd.read_csv('raw.csv')
raw_dataset.head()

_STATE  FMONTH    IDATE  IMONTH  IDAY  IYEAR  DISPCODE       SEQNO  \
0     1.0     1.0  1192021       1    19   2021    1100.0  2021000001   
1     1.0     1.0  1212021       1    21   2021    1100.0  2021000002   
2     1.0     1.0  1212021       1    21   2021    1100.0  2021000003   
3     1.0     1.0  1172021       1    17   2021    1100.0  2021000004   
4     1.0     1.0  1152021       1    15   2021    1100.0  2021000005   

           _PSU  CTELENM1  ...  _FRTRES1  _VEGRES1  _FRUTSU1  _VEGESU1  \
0  2.021000e+09       1.0  ...       1.0       1.0     100.0     214.0   
1  2.021000e+09       1.0  ...       1.0       1.0     100.0     128.0   
2  2.021000e+09       1.0  ...       1.0       1.0     100.0      71.0   
3  2.021000e+09       1.0  ...       1.0       1.0     114.0     165.0   
4  2.021000e+09       1.0  ...       1.0       1.0     100.0     258.0   

   _FRTLT1A  _VEGLT1A  _FRT16A  _VEG23A  _FRUITE1  _VEGETE1  
0       1.0       1.0      1.0      1.0       0.0       0.0  
1       1.0       1.0      1.0      1.0       0.0       0.0  
2       1.0       2.0      1.0      1.0       0.0       0.0  
3       1.0       1.0      1.0      1.0       0.0       0.0  
4       1.0       1.0      1.0      1.0       0.0       0.0  

[5 rows x 303 columns]

### Step 1: Extract columns that are relevant to our problem, i.e., hypertension analysis

After extensive secondary research, the following factors are identified to have close relationship with hypertension:
- **BMI**: The heavier we are, the more blood is required to supply oxygen and nutrients around our body, thus increasing the pressure on our blood vessels.

- **Physical Exercise**: Individuals who lead sedentary lifestyles tend to have a higher heart rate than those who are physically active. This means that the heart needs to work harder. In addition, with little physical activity, there is a high likelihood of the person being overweight.

- **Diabetes**: research has shown that people with diabetes are two to three times more likely to develop hypertension than those without diabetes. Furthermore, hypertension is a major risk factor for cardiovascular disease, which is the leading cause of death in people with diabetes. 

- **High Cholesterol**: Additionally, people with high cholesterol often develop high blood pressure. This is due to Cholesterol causing your arteries to become hard and narrow resulting in your heart straining much harder to pump blood through them. As a result, blood pressure increases.

- **Junk Food Intake**: Junk foods such as processed foods, fast foods, sugary drinks, and snacks tend to be high in salt, sugar, unhealthy fats, and calories, but low in essential nutrients such as fiber, vitamins, and minerals. Overconsumption of these foods has been linked to several health problems, including hypertension

- **Fruit Intake**: fruits are considered an important component of a healthy diet and are rich in nutrients such as vitamins, minerals, and antioxidants that are important for maintaining overall health. Some studies have suggested that increased fruit intake may be associated with a lower risk of hypertension.

- **Vegetables Intake**: Vegetables are an important part of a healthy diet, and they are rich in nutrients such as vitamins, minerals, and fiber that are essential for maintaining overall health. Several studies have suggested that increased vegetable intake may be associated with a lower risk of hypertension.

- **Education Level**: Several studies have shown that individuals with lower levels of education are more likely to develop hypertension compared to those with higher levels of education. For example, a study published in the Journal of Hypertension in 2014 found that individuals with a lower level of education had a higher risk of hypertension compared to those with a higher level of education.

- **Alcohol Consumption Habit**: Drinking too much alcohol can raise blood pressure to unhealthy levels. Having more than three drinks in one sitting temporarily increases blood pressure, but repeated binge drinking can lead to long-term increased blood pressure.

- **Smoker**: Smoking can raise our blood pressure temporarily and its chemicals can damage our artery walls, causing our arteries to narrow and stiffen.

- **Other Mental Health Problem**: Several studies have shown that individuals with mental health conditions such as depression, anxiety, and chronic stress are at a higher risk of developing hypertension compared to those without these conditions.

- **Other Physical Health Problem**: Based on multiple recent researches conduected, there are several medical conditions that are known to be risk factors for hypertension, including asthma, heart problem, arthritis.
- **Age**: As people age, their risk of developing hypertension increases. This is because the blood vessels become stiffer and less elastic over time, which can lead to an increase in blood pressure. Additionally, as people age, they may be more likely to develop other medical conditions that are known to be risk factors for hypertension, such as obesity, diabetes, and kidney disease.

- **Race**: Studies have shown that certain racial and ethnic groups, including African Americans, Hispanics, and Native Americans, are at a higher risk of developing hypertension compared to non-Hispanic whites. For example, according to the American Heart Association, African Americans have the highest prevalence of hypertension among any racial or ethnic group in the United States.


Based on the factors identified through secondary research, relevant variables and data are extracted from the raw survey data file. Variables extracted include:
1. _TOTINDA
2. _BMI5
3. DROCDY3_
4. AVEDRNK3
5. _RFBING5
6. CHOLMED3
7. FRENCHF1
8. FRUTDA2_
9. FTJUDA2_
10. VEGEDA2_
11. MENTHLTH
12. PHYSHLTH
13. _RFHYPE6
14. DIABETE4
15. _RFCHOL3

For detailed description on the variables, please refer to our data description file.

In [3]:
#Extract relevant variable columns from the raw data file
hypertension_dataset = pd.DataFrame(raw_dataset[['_TOTINDA','_BMI5','DROCDY3_','AVEDRNK3','_RFBING5','CHOLMED3','FRENCHF1','FRUTDA2_','FTJUDA2_','VEGEDA2_','MENTHLTH','PHYSHLTH','_RFHYPE6','DIABETE4','_RFCHOL3','_MICHD','_EDUCAG','_CASTHM1','_RFSMOK3','_DRDXAR3','_IMPRACE','_AGE80']])
hypertension_dataset.head()

_TOTINDA   _BMI5  DROCDY3_  AVEDRNK3  _RFBING5  CHOLMED3  FRENCHF1  \
0       2.0  1454.0       0.0       NaN       1.0       1.0     203.0   
1       1.0     NaN       0.0       NaN       1.0       1.0     555.0   
2       2.0  2829.0       0.0       NaN       1.0       2.0     201.0   
3       1.0  3347.0      14.0       3.0       2.0       2.0     204.0   
4       1.0  2873.0       0.0       NaN       1.0       1.0     202.0   

   FRUTDA2_  FTJUDA2_  VEGEDA2_  ...  _RFHYPE6  DIABETE4  _RFCHOL3  _MICHD  \
0     100.0       0.0     100.0  ...       1.0       3.0       2.0     2.0   
1     100.0       0.0     100.0  ...       2.0       1.0       2.0     1.0   
2     100.0       0.0      43.0  ...       2.0       1.0       1.0     1.0   
3      43.0      71.0      71.0  ...       2.0       1.0       2.0     2.0   
4     100.0       0.0     100.0  ...       1.0       1.0       2.0     1.0   

   _EDUCAG  _CASTHM1  _RFSMOK3  _DRDXAR3  _IMPRACE  _AGE80  
0      2.0       2.0       1.0       1.0       1.0    70.0  
1      4.0       1.0       1.0       1.0       2.0    67.0  
2      2.0       1.0       1.0       2.0       2.0    72.0  
3      2.0       1.0       1.0       2.0       1.0    62.0  
4      1.0       1.0       1.0       2.0       6.0    76.0  

[5 rows x 22 columns]

In [4]:
hypertension_dataset.shape

(438693, 22)

In [5]:
hypertension_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438693 entries, 0 to 438692
Data columns (total 22 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   _TOTINDA  438693 non-null  float64
 1   _BMI5     391841 non-null  float64
 2   DROCDY3_  438693 non-null  float64
 3   AVEDRNK3  210422 non-null  float64
 4   _RFBING5  438693 non-null  float64
 5   CHOLMED3  377122 non-null  float64
 6   FRENCHF1  400582 non-null  float64
 7   FRUTDA2_  394742 non-null  float64
 8   FTJUDA2_  394344 non-null  float64
 9   VEGEDA2_  390165 non-null  float64
 10  MENTHLTH  438691 non-null  float64
 11  PHYSHLTH  438690 non-null  float64
 12  _RFHYPE6  438693 non-null  float64
 13  DIABETE4  438690 non-null  float64
 14  _RFCHOL3  377542 non-null  float64
 15  _MICHD    434058 non-null  float64
 16  _EDUCAG   438693 non-null  float64
 17  _CASTHM1  438693 non-null  float64
 18  _RFSMOK3  438693 non-null  float64
 19  _DRDXAR3  435797 non-null  float64
 20  _IMP

### Step 2: Tackle the Missing Values

Since the raw data file is from an actual survey conducted in the US, there are missing values in many variable columns because the survey respondents failed to provide answers for these questions.

These missing values have to be tackled before we proceed with further analysis and model building. In our project, we opted for the simplest and the most direct way, which is to remove all the rows containing missing values. This method is acceptable in our case because there are `438,693` data points in the original data file and we would still have adequate amount of data for analysis after dropping the rows. 

In [6]:
#Check the number of NaN values in each variable column
hypertension_dataset.isnull().sum()

_TOTINDA         0
_BMI5        46852
DROCDY3_         0
AVEDRNK3    228271
_RFBING5         0
CHOLMED3     61571
FRENCHF1     38111
FRUTDA2_     43951
FTJUDA2_     44349
VEGEDA2_     48528
MENTHLTH         2
PHYSHLTH         3
_RFHYPE6         0
DIABETE4         3
_RFCHOL3     61151
_MICHD        4635
_EDUCAG          0
_CASTHM1         0
_RFSMOK3         0
_DRDXAR3      2896
_IMPRACE         0
_AGE80           0
dtype: int64

In [7]:
#Drop the rows with missing values
hypertension_dataset.dropna(axis=0,inplace=True)

In [8]:
#Check the number of NaN values in each variable column again
hypertension_dataset.isnull().sum()

_TOTINDA    0
_BMI5       0
DROCDY3_    0
AVEDRNK3    0
_RFBING5    0
CHOLMED3    0
FRENCHF1    0
FRUTDA2_    0
FTJUDA2_    0
VEGEDA2_    0
MENTHLTH    0
PHYSHLTH    0
_RFHYPE6    0
DIABETE4    0
_RFCHOL3    0
_MICHD      0
_EDUCAG     0
_CASTHM1    0
_RFSMOK3    0
_DRDXAR3    0
_IMPRACE    0
_AGE80      0
dtype: int64

In [9]:
#Check the number of sample points left after dropping the NaN values. 
#In this case, we still have 159,613 rows left, which is more than sufficient for further analysis and model building.
hypertension_dataset.shape

(159613, 22)

### Step 3: Tackling irrelevant data entries

Many questions in the survey are designed to include options like "Don’t know" and "Refused to tell", which allows respondents to skip a question if they are uncomfortable or unwilling to provide an answer.

However, data points with such input are not useful for our analysis and will not provide us with insighful information on the relationship between hypertension and the factors. Therefore, we chose to drop data rows with these irrelevant responses. 

For example, for the variable `_TOTINDA`, `9` indicates "Don’t know/Refused/Missing". Thus, rows with the value of 9 is dropped. 

At the same time, encoding for some variables is not intuitive enough. For instance, there are 2 numeric variable `MENTHLTH` and `PHYSHLTH`, which indicate the number of days in the last month in which the survey respondents had bad mental and physical health respectively. For these 2 variables, the value `88` indicates `None`, meaning that they do not have days with bad health. To make these numeric variables more intuitive, the numeric value `88` is replaced with `0`. 


In [10]:
#Droping rows with irrelevant survey responses. 
#Since each variale column has different encoding for "Don’t know/Refused/Missing" responses, we have to drop these values individually for each variable

#_RFHYPE6
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['_RFHYPE6'] == 9].index, inplace=True) #Don’t know/Refused/Missing

#_TOTINDA
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['_TOTINDA'] == 9].index, inplace=True) #Don’t know/Refused/Missing

#_RFBING5
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['_RFBING5'] == 9].index, inplace=True) #Don’t know/Refused/Missing

#CHOLMED3
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['CHOLMED3'] == 7].index, inplace=True) #Don’t know/Not Sure
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['CHOLMED3'] == 9].index, inplace=True) #Refused

#FRENCHF1
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['FRENCHF1'] == 777].index, inplace=True) #Don’t know/Not Sure
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['FRENCHF1'] == 999].index, inplace=True) #Refused

#MENTHLTH
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['MENTHLTH'] == 77].index, inplace=True) #Don’t know/Not sure
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['MENTHLTH'] == 99].index, inplace=True) #Refused
        
#PHYSHLTH
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['PHYSHLTH'] == 77].index, inplace=True) #Don’t know/Not sure
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['PHYSHLTH'] == 99].index, inplace=True) #Refused

for ind in hypertension_dataset.index:
    if hypertension_dataset['MENTHLTH'][ind]==88: #88 means None, so replace with 0
        hypertension_dataset['MENTHLTH'][ind] = 0
    if hypertension_dataset['PHYSHLTH'][ind]==88: #88 means None, so replace with 0
        hypertension_dataset['PHYSHLTH'][ind] = 0

#DIABETE4
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['DIABETE4'] == 9].index, inplace=True)
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['DIABETE4'] == 7].index, inplace=True)

#_RFCHOL
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['_RFCHOL3'] == 9].index, inplace=True) #Don’t know/Refused/Missing

#_EDUCAG
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['_EDUCAG'] == 9].index, inplace=True) #Don’t know/Refused/Missing

#_CASTHM1 
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['_CASTHM1'] == 9].index, inplace=True) #Don’t know/Refused/Missing

#_RFSMOK3
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['_RFSMOK3'] == 9].index, inplace=True) #Refused

#DROCDY3_
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['DROCDY3_'] == 900].index, inplace=True) #Refused

#AVEDRNK3
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['AVEDRNK3'] == 77].index, inplace=True) #Refused
hypertension_dataset.drop(hypertension_dataset[hypertension_dataset['AVEDRNK3'] == 99].index, inplace=True) #Refused


### Step 4: Create new variables by combining or transforming existing ones

The survey conducted is very comprehensive and may contain several questions related to the same factor. Thus, we can combines related factors and create a new variable that is more useful for our analysis. In addition, the scale of some numeric answer may need to be adjusted before proceeding for further exploratory data analysis.

In our project, we have created 2 new independent variables `AlchoIntake` and `FriesFreq` after merging and transforming original data columns. 

>`AlchoIntake` </br>
In the extracted dataset, 2 variables appear to be related: `DROCDY3_` refers to drink occasions per day and `AVEDRNK3` refers how many drinks the survey respondents drink on average when they drink. Thus, we can combine these 2 variables to compute their alcohol consumption on a weekly basis, which would be a more useful numeric variable for analysis and model building.

>`FriesFreq`<br/>
`FRENCHF1` indicates the frequency of survey respondents eating any kind of fried potatoes, including french fries, home fries, or hash browns. However, this numeric value have different unit for different ranges, e.g., 101 - 199 is in days while 201 - 299 is in weeks. Thus, we have to standardise the units to days and calculate the respondents' average frequency of consuming french fries.


The detailed manipulations can be found in the programs below.

In [11]:
#Combine DROCDY3_ and AVEDRNK3 2 to compute alcohol consumption on a weekly basis and store it in a new variable called AlchoIntake

#Extract varaibles and create a new empty column to store the result
Alcho_Consumption = hypertension_dataset[['DROCDY3_','AVEDRNK3']]
Alcho_Consumption = Alcho_Consumption.assign(AlchoIntake = " ")

#Compute weekly alcohol consumption of survey respondents, the logic is drived based on codebook provided by the original survey
for ind in Alcho_Consumption.index:    
    if Alcho_Consumption['DROCDY3_'][ind]==0 or Alcho_Consumption['AVEDRNK3'][ind] == 88.0: #means no alcohol consumption
        Alcho_Consumption['AlchoIntake'][ind] = 0
    else: 
        Alcho_Consumption['AlchoIntake'][ind] = Alcho_Consumption['DROCDY3_'][ind] * Alcho_Consumption['AVEDRNK3'][ind] * 7 /100 #compute weekly alcohol consumption 

#Convert the data type from object to float
Alcho_Consumption['AlchoIntake'] = Alcho_Consumption['AlchoIntake'].astype(float)

#Concatenate the new variable column with the current dataset
hypertension_dataset = pd.concat([hypertension_dataset,Alcho_Consumption['AlchoIntake']],axis=1).reindex(hypertension_dataset.index)


/var/folders/rq/pmx086v15tqccq0vw0tdmn200000gn/T/ipykernel_19283/3835171372.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Alcho_Consumption['AlchoIntake'][ind] = Alcho_Consumption['DROCDY3_'][ind] * Alcho_Consumption['AVEDRNK3'][ind] * 7 /100 #compute weekly alcohol consumption
/var/folders/rq/pmx086v15tqccq0vw0tdmn200000gn/T/ipykernel_19283/3835171372.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Alcho_Consumption['AlchoIntake'][ind] = 0


In [12]:
#Standardise the unit of measurement for FRENCHF1 to days and store in a new vairable called FriesFreq

#Extract varaibles and create a new empty column to store the result
Fren_Fries = pd.DataFrame(hypertension_dataset['FRENCHF1'])
Fren_Fries = Fren_Fries.assign(FriesFreq = " ")

#Convert unit of measurement to days, the logic is drived based on codebook provided by the original survey
for ind in Fren_Fries.index:    
    if 100 < Fren_Fries['FRENCHF1'][ind] < 200 : #original unit is in days
        Fren_Fries['FriesFreq'][ind] = Fren_Fries['FRENCHF1'][ind] - 100
        
    elif 200 < Fren_Fries['FRENCHF1'][ind] < 300 :  #original unit in weeks
        Fren_Fries['FriesFreq'][ind] = (Fren_Fries['FRENCHF1'][ind] - 200)/7
    
    elif Fren_Fries['FRENCHF1'][ind] == 300 :  #less than once a month
        Fren_Fries['FriesFreq'][ind] = 0.02  #just set to 0.02 which is lower than 1/30
        
    elif 300 < Fren_Fries['FRENCHF1'][ind] < 400 :  #original unit in month
        Fren_Fries['FriesFreq'][ind] = (Fren_Fries['FRENCHF1'][ind] - 300)/30
        
    elif Fren_Fries['FRENCHF1'][ind] == 555 :  #never eat 
        Fren_Fries['FriesFreq'][ind] = 0 

#Convert the data type from object to float
Fren_Fries['FriesFreq'] = Fren_Fries['FriesFreq'].astype(float)

#Concatenate the new variable column with the current dataset
hypertension_dataset = pd.concat([hypertension_dataset,Fren_Fries['FriesFreq']],axis=1).reindex(hypertension_dataset.index)

/var/folders/rq/pmx086v15tqccq0vw0tdmn200000gn/T/ipykernel_19283/1139661966.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Fren_Fries['FriesFreq'][ind] = (Fren_Fries['FRENCHF1'][ind] - 200)/7
/var/folders/rq/pmx086v15tqccq0vw0tdmn200000gn/T/ipykernel_19283/1139661966.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Fren_Fries['FriesFreq'][ind] = (Fren_Fries['FRENCHF1'][ind] - 300)/30
/var/folders/rq/pmx086v15tqccq0vw0tdmn200000gn/T/ipykernel_19283/1139661966.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.p

In [13]:
hypertension_dataset.head()

_TOTINDA   _BMI5  DROCDY3_  AVEDRNK3  _RFBING5  CHOLMED3  FRENCHF1  \
3        1.0  3347.0      14.0       3.0       2.0       2.0     204.0   
9        1.0  3994.0      10.0       2.0       1.0       1.0     202.0   
19       2.0  2986.0      29.0       1.0       1.0       1.0     302.0   
22       2.0  3587.0      13.0       2.0       1.0       1.0     304.0   
28       1.0  2905.0     100.0       1.0       1.0       2.0     202.0   

    FRUTDA2_  FTJUDA2_  VEGEDA2_  ...  _RFCHOL3  _MICHD  _EDUCAG  _CASTHM1  \
3       43.0      71.0      71.0  ...       2.0     2.0      2.0       1.0   
9      100.0       0.0      57.0  ...       1.0     2.0      2.0       1.0   
19      29.0     200.0     100.0  ...       2.0     2.0      2.0       2.0   
22       7.0       3.0      20.0  ...       2.0     2.0      2.0       1.0   
28     100.0     100.0     100.0  ...       2.0     2.0      3.0       1.0   

    _RFSMOK3  _DRDXAR3  _IMPRACE  _AGE80  AlchoIntake  FriesFreq  
3        1.0       2.0       1.0    62.0         2.94   0.571429  
9        1.0       1.0       1.0    65.0         1.40   0.285714  
19       1.0       2.0       1.0    53.0         2.03   0.066667  
22       2.0       2.0       1.0    68.0         1.82   0.133333  
28       1.0       1.0       1.0    80.0         7.00   0.285714  

[5 rows x 24 columns]

### Step 5: Decode categorical variables based on data description

In the extracted data file, survey responses for categorical variables have been encoded using different numeric values, which explains why the data type for all columns are float64. However, the encoded number does not convey the original meaning of each categories and we have to refer to the codebook to interpret the value. Thus, we decided to decode the categorical variables based on our data documentation, so that it is more intuitive and esaier for human interpretation. 

Categorical variables include:
1. _TOTINDA
2. _RFBING5
3. CHOLMED3
4. _RFHYPE6
5. DIABETE4
6. _RFCHOL3
7. _MICHD
8. _EDUCAG
9. _CASTHM1
10. _RFSMOK3
11. _DRDXAR3
12. _IMPRACE
13. _RFDRHV7

In [14]:
#Decode the categorical variable 
#This has to be done manually because the encoding is not consistent, i.e., 1.0 is "yes" for some variables but "no" for others
for ind in hypertension_dataset.index:    
    #Decode _RFHYPE6 to be yes/no
    if hypertension_dataset['_RFHYPE6'][ind]==1.0:
        hypertension_dataset['_RFHYPE6'][ind] = "No"
    elif hypertension_dataset['_RFHYPE6'][ind]==2.0:
        hypertension_dataset['_RFHYPE6'][ind] = "Yes"
    
    #Decode _TOTINDA
    if hypertension_dataset['_TOTINDA'][ind]==1.0:
        hypertension_dataset['_TOTINDA'][ind] = "Yes"
    elif hypertension_dataset['_TOTINDA'][ind]==2.0:
        hypertension_dataset['_TOTINDA'][ind] = "No"
        
    #Decode _RFBING5
    if hypertension_dataset['_RFBING5'][ind]==1.0:
        hypertension_dataset['_RFBING5'][ind] = "No"
    elif hypertension_dataset['_RFBING5'][ind]==2.0:
        hypertension_dataset['_RFBING5'][ind] = "Yes"
        
    #Decode CHOLMED3
    if hypertension_dataset['CHOLMED3'][ind]==1.0:
        hypertension_dataset['CHOLMED3'][ind] = "Yes"
    elif hypertension_dataset['CHOLMED3'][ind]==2.0:
        hypertension_dataset['CHOLMED3'][ind] = "No"
        
    #Decode diabetes_dataset to be yes/no
    if hypertension_dataset['DIABETE4'][ind]==1.0:
        hypertension_dataset['DIABETE4'][ind] = "Yes"
    elif hypertension_dataset['DIABETE4'][ind]==2.0:
        hypertension_dataset['DIABETE4'][ind] = "Yes"
    elif hypertension_dataset['DIABETE4'][ind]==3.0:
        hypertension_dataset['DIABETE4'][ind] = "No"
    elif hypertension_dataset['DIABETE4'][ind]==4.0:
        hypertension_dataset['DIABETE4'][ind] = "No"
 
    #Decode _RFCHOL3 to be yes/no
    if hypertension_dataset['_RFCHOL3'][ind]==1.0:
        hypertension_dataset['_RFCHOL3'][ind] = "No"
    elif hypertension_dataset['_RFCHOL3'][ind]==2.0:
        hypertension_dataset['_RFCHOL3'][ind] = "Yes"
    
    #Decode _MICHD to be yes/no
    if hypertension_dataset['_MICHD'][ind]==1.0:
        hypertension_dataset['_MICHD'][ind] = "Yes"
    elif hypertension_dataset['_MICHD'][ind]==2.0:
        hypertension_dataset['_MICHD'][ind] = "No"
        
    #Decode _EDUCAG 
    if hypertension_dataset['_EDUCAG'][ind]==1.0:
        hypertension_dataset['_EDUCAG'][ind] = "Did not graduate High School"
    elif hypertension_dataset['_EDUCAG'][ind]==2.0:
        hypertension_dataset['_EDUCAG'][ind] = "Graduated High School"
    elif hypertension_dataset['_EDUCAG'][ind]==3.0:
        hypertension_dataset['_EDUCAG'][ind] = "Attended College or Technical School"
    elif hypertension_dataset['_EDUCAG'][ind]==4.0:
        hypertension_dataset['_EDUCAG'][ind] = "Graduated from College or Technical School" 
    
    #Decode _CASTHM1 
    if hypertension_dataset['_CASTHM1'][ind]==1.0:
        hypertension_dataset['_CASTHM1'][ind] = "No"
    elif hypertension_dataset['_CASTHM1'][ind]==2.0:
        hypertension_dataset['_CASTHM1'][ind] = "Yes"
        
    #Decode _RFSMOK3
    if hypertension_dataset['_RFSMOK3'][ind]==1.0:
        hypertension_dataset['_RFSMOK3'][ind] = "No"
    elif hypertension_dataset['_RFSMOK3'][ind]==2.0:
        hypertension_dataset['_RFSMOK3'][ind] = "Yes"
        
    #Decode _DRDXAR3
    if hypertension_dataset['_DRDXAR3'][ind]==1.0:
        hypertension_dataset['_DRDXAR3'][ind] = "Yes"
    elif hypertension_dataset['_DRDXAR3'][ind]==2.0:
        hypertension_dataset['_DRDXAR3'][ind] = "No"

    #Decode _IMPRACE 
    if hypertension_dataset['_IMPRACE'][ind]==1.0:
        hypertension_dataset['_IMPRACE'][ind] = "White"
    elif hypertension_dataset['_IMPRACE'][ind]==2.0:
        hypertension_dataset['_IMPRACE'][ind] = "Black"
    elif hypertension_dataset['_IMPRACE'][ind]==3.0:
        hypertension_dataset['_IMPRACE'][ind] = "Asian"
    elif hypertension_dataset['_IMPRACE'][ind]==4.0:
        hypertension_dataset['_IMPRACE'][ind] = "American Indian" 
    elif hypertension_dataset['_IMPRACE'][ind]==5.0:
        hypertension_dataset['_IMPRACE'][ind] = "Hispanic" 
    elif hypertension_dataset['_IMPRACE'][ind]==6.0:
        hypertension_dataset['_IMPRACE'][ind] = "Other Race" 
    
        

/var/folders/rq/pmx086v15tqccq0vw0tdmn200000gn/T/ipykernel_19283/3621867782.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hypertension_dataset['_TOTINDA'][ind] = "Yes"
/Users/wrx/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)
/var/folders/rq/pmx086v15tqccq0vw0tdmn200000gn/T/ipykernel_19283/3621867782.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/inde

In [15]:
hypertension_dataset.shape

(149931, 24)

In [16]:
hypertension_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 149931 entries, 3 to 438692
Data columns (total 24 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   _TOTINDA     149931 non-null  object 
 1   _BMI5        149931 non-null  float64
 2   DROCDY3_     149931 non-null  float64
 3   AVEDRNK3     149931 non-null  float64
 4   _RFBING5     149931 non-null  object 
 5   CHOLMED3     149931 non-null  object 
 6   FRENCHF1     149931 non-null  float64
 7   FRUTDA2_     149931 non-null  float64
 8   FTJUDA2_     149931 non-null  float64
 9   VEGEDA2_     149931 non-null  float64
 10  MENTHLTH     149931 non-null  float64
 11  PHYSHLTH     149931 non-null  float64
 12  _RFHYPE6     149931 non-null  object 
 13  DIABETE4     149931 non-null  object 
 14  _RFCHOL3     149931 non-null  object 
 15  _MICHD       149931 non-null  object 
 16  _EDUCAG      149931 non-null  object 
 17  _CASTHM1     149931 non-null  object 
 18  _RFSMOK3     149931 non-

### Step 6: Compute true values for numeric variables

For some numeric data, their actual values are multipled by `100` before being stored into the data file, this is to prevent potential truncation and data loss. Thus, these variables have `2 implied decimal places` according to the survey codebook, e.g., `9999` actually means `99.99`. To get the actual values, we divided them by `100`. 

Numeric variables that need adjustment include:
1. _BMI5
2. FRUTDA2_
3. FTJUDA2_
4. VEGEDA2_

In [17]:
#divide "implied 2 dp" variables by 100
hypertension_dataset['_BMI5'] = hypertension_dataset['_BMI5'].div(100).round(2)
hypertension_dataset['FRUTDA2_'] = hypertension_dataset['FRUTDA2_'].div(100).round(2)
hypertension_dataset['FTJUDA2_'] = hypertension_dataset['FTJUDA2_'].div(100).round(2)
hypertension_dataset['VEGEDA2_'] = hypertension_dataset['VEGEDA2_'].div(100).round(2)

### Step 7: Identify and remove outliers for numeric variables

For numeric variables, it is essential to check the existence of outliers. In our case, since the survey are collected from US citizens, there may be situations where the survey respondents gives inaccurate or fake data, causing outliers to exist. These Outliers can affect data visualization, such as scatter plots or histograms to be skewed, making it difficult to interpret the data. At the same time, outliers can have a disproportionate impact on predictive models, leading to inaccurate predictions. 

Conducing outliers analysis allows us to identify errors and remove anomaliers in the data before proceeding with further exploratory data analysis.

In [18]:
#Check number of outliers for each numeric variable
def outliers_counter(df):
        Q1=df.quantile(0.25)
        Q3=df.quantile(0.75)
        IQR=Q3-Q1
        
        count=0
        for i in df:
            if i<(Q1-1.5*IQR) or i>(Q3+1.5*IQR):
                count+=1
        
        return count
    
print("_BMI5 has",outliers_counter(hypertension_dataset["_BMI5"]),"outliers.")
print("FriesFreq has",outliers_counter(hypertension_dataset["FriesFreq"]),"outliers.")
print("FRUTDA2_ has",outliers_counter(hypertension_dataset["FRUTDA2_"]),"outliers.")
print("FTJUDA2_ has",outliers_counter(hypertension_dataset["FTJUDA2_"]),"outliers.")
print("VEGEDA2_ has",outliers_counter(hypertension_dataset["VEGEDA2_"]),"outliers.")
print("AlchoIntake has",outliers_counter(hypertension_dataset["AlchoIntake"]),"outliers.")

_BMI5 has 4662 outliers.
FriesFreq has 8523 outliers.
FRUTDA2_ has 35992 outliers.
FTJUDA2_ has 21025 outliers.
VEGEDA2_ has 23032 outliers.
AlchoIntake has 9780 outliers.


In [19]:
#Determine the index of rows with outliers
def outliers_index(df):
        Q1=df.quantile(0.25)
        Q3=df.quantile(0.75)
        IQR=Q3-Q1
        
        index_list=[]
        index=0
        for i in df:
            if i<(Q1-1.5*IQR) or i>(Q3+1.5*IQR):
                index_list.append(index)
            index+=1
        return index_list

#Remove rows with outliers
numeric_vars = ['FTJUDA2_','_BMI5','FRUTDA2_','VEGEDA2_','FriesFreq','AlchoIntake']
for var in numeric_vars:
    if outliers_counter(hypertension_dataset[var])>0:
        index_to_drop=outliers_index(hypertension_dataset[var])
        hypertension_dataset = hypertension_dataset.drop(hypertension_dataset.index[index_to_drop])

In [20]:
#Since the removal of outliers is done in phase, new outliers may be created in some variables after rows are removed in other variables
#Nevertheless, the number of outliers are reduced significantly.
print("_BMI5 has",outliers_counter(hypertension_dataset["_BMI5"]),"outliers.")
print("FRNCHDA_ has",outliers_counter(hypertension_dataset["FriesFreq"]),"outliers.")
print("FRUTDA2_ has",outliers_counter(hypertension_dataset["FRUTDA2_"]),"outliers.")
print("FTJUDA2_ has",outliers_counter(hypertension_dataset["FTJUDA2_"]),"outliers.")
print("VEGEDA2_ has",outliers_counter(hypertension_dataset["VEGEDA2_"]),"outliers.")
print("AlchoIntake has",outliers_counter(hypertension_dataset["AlchoIntake"]),"outliers.")

_BMI5 has 323 outliers.
FRNCHDA_ has 0 outliers.
FRUTDA2_ has 0 outliers.
FTJUDA2_ has 8120 outliers.
VEGEDA2_ has 0 outliers.
AlchoIntake has 4169 outliers.


### Cleaned Dataset

After the data cleaning process, we have produced a cleaned dataset with `76,825` rows, which is still considered as adequate for subsequent data analysis and model building. 

This cleaned dataset is then exported to a new csv file called hypertension.

In [21]:
hypertension_dataset.shape

(76825, 24)

In [22]:
#Output cleaned dataset
hypertension_dataset.to_csv("hypertension.csv")